File: 22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python3
#
# [22] 括号生成
#

# @lc code=start
from typing import List
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        tmp = ''
        self.generate(1, 0, res, tmp+'(', n)
        return res


    def generate(self, left: int, right: int, res: List[str], tmp: str, n: int):
        if left == right and left == n:
            res.append(tmp)
            return
        if left < n:
            self.generate(left+1, right, res, tmp+'(', n)
        if left > right:
            self.generate(left, right+1, res, tmp+')', n)

# @lc code=end

File: 22.括号生成.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=22 lang=java
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        StringBuffer tmp = new StringBuffer();
        generate(1, 0, res, tmp.append('('), n);
        return res;
        
    }
    public void generate(int left, int right, List<String> res, StringBuffer tmp, int n) {
        if(left==right && left == n){
            res.add(tmp.toString());
            return;
        }
        if(left < n){
            generate(left+1, right, res, tmp.append('('), n);
            tmp.deleteCharAt(tmp.length()-1);
        }
        if(left>right){
            generate(left, right+1, res, tmp.append(')'), n);
            tmp.deleteCharAt(tmp.length()-1);
        }
    }
}
// @lc code=end

